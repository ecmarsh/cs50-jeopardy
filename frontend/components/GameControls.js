import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Link from 'next/link';
import Router from 'next/router';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import GameReset from './GameReset';
import Button from './styles/Button';
import { ThemeToggler } from './Theme';
import SaveGameButton from './StudyGameCreate';
import { Slide } from './Animated';

const LOCAL_RESET_MUTATION = gql`
  mutation {
    resetBoard @client
  }
`;

function ActionBar({ gameName, round, config, isOwner }) {
  const [hovered, toggle] = useState(false);

  return (
    <Slide
      className="left"
      onMouseEnter={() => toggle(true)}
      onMouseLeave={() => toggle(false)}
      hovered={hovered}
    >
      <div className="inner" data-test="game-control-bar">
        {isOwner && (
          <>
            <Link
              href={{
                pathname: 'launch',
                query: { game: gameName },
              }}
            >
              <a> ◁ Back to Edit Game</a>
            </Link>

            <GameReset
              gameName={gameName}
              hasDoubleJeopardy={config.hasDoubleJeopardy}
            >
              {resetGameMutation => (
                <Button
                  tertiary
                  onClick={async e => {
                    if (confirm('Are you sure you want to reset your game?')) {
                      e.preventDefault();
                      await resetGameMutation();
                      Router.push({
                        pathname: `/play`,
                        query: { game: gameName },
                      });
                    }
                  }}
                >
                  Reset Game
                </Button>
              )}
            </GameReset>
          </>
        )}
        {!isOwner && (
          <>
            <Mutation mutation={LOCAL_RESET_MUTATION}>
              {resetBoard => (
                <Button
                  tertiary
                  onClick={e => {
                    e.preventDefault();
                    if (confirm('Are you sure you want to reset this game?')) {
                      resetBoard();
                      // route back to first round
                      Router.push({
                        pathname: `/play`,
                        query: { game: gameName },
                      });
                    }
                  }}
                >
                  Reset Game
                </Button>
              )}
            </Mutation>
          </>
        )}
        <ThemeToggler buttonTxt={`Toggle Theme`} />
        {round <= (isOwner ? 2 : 1) && (
          <Link
            href={{
              pathname: 'play',
              query: { game: gameName, round: round + 1 },
            }}
          >
            <a> Next Round ▷</a>
          </Link>
        )}
        {!isOwner && (
          <SaveGameButton gameName={gameName}>Finish Game ✓</SaveGameButton>
        )}
      </div>
      <div className={'bar-icon'}>
        <img src="../static/settings-icon.png" alt="game-control-icon" />
      </div>
    </Slide>
  );
}

ActionBar.propTypes = {
  gameName: PropTypes.string.isRequired,
  round: PropTypes.number.isRequired,
  config: PropTypes.object.isRequired,
  isOwner: PropTypes.bool.isRequired,
};

ActionBar.defaultProps = {
  round: 1,
  isOwner: false,
};

export default ActionBar;
