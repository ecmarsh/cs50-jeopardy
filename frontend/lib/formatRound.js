export default function(page) {
  let title;
  switch (page) {
    case 1:
      title = 'Single';
      break;
    case 2:
      title = 'Double';
      break;
    case 3:
      title = 'Final';
      break;
    default:
      title = 'CS50';
  }

  return title + ' Jeopardy';
}
