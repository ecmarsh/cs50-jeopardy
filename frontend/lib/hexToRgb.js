export default function(hex) {
  const chunk = 2,
    radix = 16,
    rgb = [];

  let i = 1;
  while (hex && i < hex.length) {
    rgb.push(parseInt(hex.substring(i, i + chunk), radix));
    i += chunk;
  }

  return rgb.toString();
}
