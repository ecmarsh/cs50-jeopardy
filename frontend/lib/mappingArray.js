export default function(N = 0, isFilled = true, startAtZero = false) {
  if (N == 0) return [];
  if (N > 0 && !isFilled) return Array(N).fill();
  if (isFilled && !startAtZero)
    return Array.from({ length: N }, (v, k) => k + 1);
  return Array.from({ length: N }, (v, k) => k++);
}
