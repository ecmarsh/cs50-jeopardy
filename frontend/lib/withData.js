import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint, endpointProd } from '../config';
import { LOCAL_STATE_QUERY } from '../components/SoundEffect';
import { ANSWERED_QUESTIONS_QUERY } from '../components/GameScreen';
import gql from 'graphql-tag';

// Setup for SSR w/ next and config endpoint
function createClient({ headers }) {
  const typeDefs = gql`
    type StudyQuestion {
      data: Json
    }
  `;
  const client = new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpointProd,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    typeDefs,
    // cache
    clientState: {
      resolvers: {
        Mutation: {
          toggleSound(_, variables, { cache }) {
            // read the soundOn value from the cache
            const { soundOn } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });
            // Toggle sound on || off
            const data = {
              data: { soundOn: !soundOn },
            };
            cache.writeData(data);
            return data;
          },
          addAnswered(_, { id, val }, { cache }) {
            // read the answered questions from the cache
            const { answeredQuestions } = cache.readQuery({
              query: ANSWERED_QUESTIONS_QUERY,
            });
            const newQuestion = { id, val, __typename: 'StudyQuestion' };
            const jsonified = JSON.stringify([
              newQuestion,
              ...answeredQuestions,
            ]);
            // Add new question to data
            const data = {
              data: {
                answeredQuestions: [newQuestion, ...answeredQuestions],
              },
            };

            cache.writeData(data);
            return data;
          },
          resetBoard(_, variables, { cache }) {
            // Reset answered questions
            const data = {
              data: {
                answeredQuestions: [],
              },
            };
            cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        soundOn: false,
        answeredQuestions: [],
      },
    },
  });
  return client;
}

export default withApollo(createClient);
