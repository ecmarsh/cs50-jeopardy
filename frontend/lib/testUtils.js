import casual from 'casual';
import mappingArray from './mappingArray';

// seed it for repeatable random sequence
casual.seed(123);

const fakeQuestion = () => ({
  id: 'abc123',
  question: 'Time complexity of merge sort?',
  answer: 'n log n',
  user: {
    id: casual.uuid,
    name: 'User Name',
    __typename: 'User',
  },
  createdAt: '2019-02 - 26T19: 24: 16.000Z',
  game: {
    id: 'game123',
    name: 'GameForGeekz',
    user: {
      id: casual.uuid,
      __typename: 'User',
    },
    isPublic: casual.coin_flip,
    __typename: 'Game',
  },
  categoryQuestions: [
    {
      id: 'catQuest123',
      category: { id: 'cat123', __typename: 'Category' },
      __typename: '__CategoryQuestion',
    },
  ],
  __typename: 'Question',
});

const fakeUser = () => ({
  __typename: 'User',
  id: 'userABC123',
  name: casual.name,
  email: casual.email,
  permissions: ['USER'],
});

const fakeAdmin = () => ({
  __typename: 'User',
  id: 'adminABC123',
  name: casual.name,
  email: casual.email,
  permission: ['ADMIN', 'USER'],
  games: [],
});

const fakeCategoryQuestion = (id = 'catQuest123', difficulty = 1) => ({
  id,
  difficulty,
  isDouble: false,
  answered: casual.coin_flip,
  question: fakeQuestion(),
  category: { id: 'cat123', name: casual.word, __typename: 'Category' },
  game: { id: 'g@m3', name: 'Test Game', __typename: 'Game' },
  __typename: 'CategoryQuestion',
});

const fillCategoryQuestions = (n = 5) =>
  mappingArray(n).map(i => fakeCategoryQuestion(`catQuest-${i}`, i));

const fakeCategory = (id = 'cat123') => ({
  id,
  name: casual.word,
  game: { id: 'g@m3', name: 'Test Game' },
  categoryQuestions: fillCategoryQuestions(),
  __typename: 'Category',
});

const fillCategories = (n = 11) =>
  mappingArray(n).map(i => fakeCategory(`cat-${i}`));

const fakeTeam = id => ({
  __typename: 'Team',
  id: `t${id}`,
  name: `Team${id}`,
  score: 1000,
  game: { id: 'g@m3', name: 'Test Game' },
});

const fakeGameConfig = () => ({
  id: 'config123',
  hasDoubleJeopardy: false,
  hasRoundTimer: false,
  roundTime: 0,
  finalTime: 0,
  __typename: 'GameConfig',
});

const fakeGame = () => ({
  __typename: 'Game',
  id: 'g@m3',
  name: 'Test Game',
  createdAt: '2019-03 - 06T19: 12: 16.000Z',
  isPublic: casual.coin_flip,
  user: fakeUser(),
  teams: [fakeTeam(1), fakeTeam(2)],
  categories: fillCategories(),
  questions: [fakeQuestion()],
  config: fakeGameConfig(),
});

const fakeStudyGame = (id = 'SG123') => ({
  __typename: 'StudyGame',
  id,
  answeredQuestions: [fakeCategoryQuestion(), fakeCategoryQuestion()],
  score: 1000,
  game: fakeGame(),
  user: fakeUser(),
  createdAt: '2019-04 - 25T19: 10: 16.000Z',
});

const fakeGameSummary = () => ({
  questionsConnection: {
    aggregate: {
      count: casual.double + 1,
      __typename: 'aggregate',
    },
    __typename: 'questionsConnection',
  },
  categoryQuestionsConnection: {
    aggregate: {
      count: casual.double,
      __typename: 'aggregate',
    },
    __typename: 'categoryQuestionsConnection',
  },
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  fakeQuestion,
  fakeUser,
  fakeAdmin,
  fakeCategory,
  fakeCategoryQuestion,
  fakeTeam,
  fakeGame,
  fakeStudyGame,
  fakeGameSummary,
  fakeGameConfig,
};
