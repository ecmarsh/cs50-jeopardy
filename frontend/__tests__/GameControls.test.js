import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import GameControls from '../components/GameControls';
import { fakeGame, fakeGameConfig } from '../lib/testUtils';

const game = fakeGame();
const config = fakeGameConfig();

describe('<GameControls />', () => {
  it('renders game controls and matches snapshot for owner game', async () => {
    const wrapper = mount(
      <MockedProvider>
        <GameControls
          gameName={game.name}
          round={1}
          isOwner={true}
          config={config}
        />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const ownerGameControls = wrapper.find('div[data-test="game-control-bar"]');
    expect(toJSON(ownerGameControls)).toMatchSnapshot();
  });
});
