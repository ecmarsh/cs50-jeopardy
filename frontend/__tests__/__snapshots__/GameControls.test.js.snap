// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameControls /> renders game controls and matches snapshot for owner game 1`] = `
<div
  className="inner"
  data-test="game-control-bar"
>
  <Link
    href={
      Object {
        "pathname": "launch",
        "query": Object {
          "game": "Test Game",
        },
      }
    }
  >
    <a
      href="launch?game=Test%20Game"
      onClick={[Function]}
    >
       ‚óÅ Back to Edit Game
    </a>
  </Link>
  <GameReset
    gameName="Test Game"
    hasDoubleJeopardy={false}
  >
    <Mutation
      mutation={
        Object {
          "definitions": Array [
            Object {
              "directives": Array [],
              "kind": "OperationDefinition",
              "name": Object {
                "kind": "Name",
                "value": "GAME_RESET_MUTATION",
              },
              "operation": "mutation",
              "selectionSet": Object {
                "kind": "SelectionSet",
                "selections": Array [
                  Object {
                    "alias": undefined,
                    "arguments": Array [
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "name",
                        },
                        "value": Object {
                          "kind": "Variable",
                          "name": Object {
                            "kind": "Name",
                            "value": "gameName",
                          },
                        },
                      },
                    ],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "resetGame",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "id",
                          },
                          "selectionSet": undefined,
                        },
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "teams",
                          },
                          "selectionSet": Object {
                            "kind": "SelectionSet",
                            "selections": Array [
                              Object {
                                "alias": undefined,
                                "arguments": Array [],
                                "directives": Array [],
                                "kind": "Field",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "id",
                                },
                                "selectionSet": undefined,
                              },
                              Object {
                                "alias": undefined,
                                "arguments": Array [],
                                "directives": Array [],
                                "kind": "Field",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "name",
                                },
                                "selectionSet": undefined,
                              },
                              Object {
                                "alias": undefined,
                                "arguments": Array [],
                                "directives": Array [],
                                "kind": "Field",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "score",
                                },
                                "selectionSet": undefined,
                              },
                            ],
                          },
                        },
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "categories",
                          },
                          "selectionSet": Object {
                            "kind": "SelectionSet",
                            "selections": Array [
                              Object {
                                "alias": undefined,
                                "arguments": Array [],
                                "directives": Array [],
                                "kind": "Field",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "id",
                                },
                                "selectionSet": undefined,
                              },
                              Object {
                                "alias": undefined,
                                "arguments": Array [],
                                "directives": Array [],
                                "kind": "Field",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "categoryQuestions",
                                },
                                "selectionSet": Object {
                                  "kind": "SelectionSet",
                                  "selections": Array [
                                    Object {
                                      "alias": undefined,
                                      "arguments": Array [],
                                      "directives": Array [],
                                      "kind": "Field",
                                      "name": Object {
                                        "kind": "Name",
                                        "value": "id",
                                      },
                                      "selectionSet": undefined,
                                    },
                                    Object {
                                      "alias": undefined,
                                      "arguments": Array [],
                                      "directives": Array [],
                                      "kind": "Field",
                                      "name": Object {
                                        "kind": "Name",
                                        "value": "answered",
                                      },
                                      "selectionSet": undefined,
                                    },
                                  ],
                                },
                              },
                            ],
                          },
                        },
                      ],
                    },
                  },
                  Object {
                    "alias": undefined,
                    "arguments": Array [
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "gameName",
                        },
                        "value": Object {
                          "kind": "Variable",
                          "name": Object {
                            "kind": "Name",
                            "value": "gameName",
                          },
                        },
                      },
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "hasDoubleJeopardy",
                        },
                        "value": Object {
                          "kind": "Variable",
                          "name": Object {
                            "kind": "Name",
                            "value": "hasDoubleJeopardy",
                          },
                        },
                      },
                    ],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "updateDoubleJeopardy",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "id",
                          },
                          "selectionSet": undefined,
                        },
                      ],
                    },
                  },
                  Object {
                    "alias": undefined,
                    "arguments": Array [],
                    "directives": Array [
                      Object {
                        "arguments": Array [],
                        "kind": "Directive",
                        "name": Object {
                          "kind": "Name",
                          "value": "client",
                        },
                      },
                    ],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "resetBoard",
                    },
                    "selectionSet": undefined,
                  },
                ],
              },
              "variableDefinitions": Array [
                Object {
                  "defaultValue": undefined,
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NonNullType",
                    "type": Object {
                      "kind": "NamedType",
                      "name": Object {
                        "kind": "Name",
                        "value": "String",
                      },
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "gameName",
                    },
                  },
                },
                Object {
                  "defaultValue": undefined,
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "Boolean",
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "hasDoubleJeopardy",
                    },
                  },
                },
              ],
            },
          ],
          "kind": "Document",
          "loc": Object {
            "end": 456,
            "start": 0,
          },
        }
      }
      update={[Function]}
      variables={
        Object {
          "gameName": "Test Game",
          "hasDoubleJeopardy": false,
        }
      }
    >
      <Button
        onClick={[Function]}
        tertiary={true}
      >
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "Button-sc-1f1w3gq-0",
                "isStatic": false,
                "lastClassName": "kpEvwZ",
                "rules": Array [
                  "font-family:",
                  [Function],
                  ";color:white;border-radius:0.5rem;text-transform:uppercase;letter-spacing:1px;transition:.2s ease;&:hover{cursor:pointer;}&:focus{cursor:pointer;outline:",
                  [Function],
                  " auto 3px;}",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " &.button--sound{font-weight:bolder;text-decoration:line-through;&::after{content:\\"‚ùå\\";padding-left:3px;}",
                  [Function],
                  "}",
                  [Function],
                ],
              },
              "displayName": "Button",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "Button-sc-1f1w3gq-0",
              "target": "button",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          onClick={[Function]}
          tertiary={true}
        >
          <button
            className="Button-sc-1f1w3gq-0 kpEvwZ"
            onClick={[Function]}
          >
            Reset Game
          </button>
        </StyledComponent>
      </Button>
    </Mutation>
  </GameReset>
  <ThemeToggler
    buttonTxt="Toggle Theme"
  >
    <div
      className="current-theme"
    >
      <Button
        buttonTxt="Toggle Theme"
        onClick={[Function]}
        tertiary={true}
      >
        <StyledComponent
          buttonTxt="Toggle Theme"
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "Button-sc-1f1w3gq-0",
                "isStatic": false,
                "lastClassName": "kpEvwZ",
                "rules": Array [
                  "font-family:",
                  [Function],
                  ";color:white;border-radius:0.5rem;text-transform:uppercase;letter-spacing:1px;transition:.2s ease;&:hover{cursor:pointer;}&:focus{cursor:pointer;outline:",
                  [Function],
                  " auto 3px;}",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " &.button--sound{font-weight:bolder;text-decoration:line-through;&::after{content:\\"‚ùå\\";padding-left:3px;}",
                  [Function],
                  "}",
                  [Function],
                ],
              },
              "displayName": "Button",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "Button-sc-1f1w3gq-0",
              "target": "button",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          onClick={[Function]}
          tertiary={true}
        >
          <button
            className="Button-sc-1f1w3gq-0 kpEvwZ"
            onClick={[Function]}
          >
            Toggle Theme
          </button>
        </StyledComponent>
      </Button>
    </div>
  </ThemeToggler>
  <Link
    href={
      Object {
        "pathname": "play",
        "query": Object {
          "game": "Test Game",
          "round": 2,
        },
      }
    }
  >
    <a
      href="play?game=Test%20Game&round=2"
      onClick={[Function]}
    >
       Next Round ‚ñ∑
    </a>
  </Link>
</div>
`;
